#LGCS 124 S22
##Homework #2

Submit your response on Sakai as a text document.

IMPORTANT: In your response, be sure to include not only your results/answers but also how you got to the answer (you may submit code, but it is not required except when explicitly called for).

In [ ]:
# Drive mounting and downloading cell

from google.colab import drive
drive.mount('/content/drive')

import nltk
nltk.download("punkt")
nltk.download("stopwords")
from nltk.corpus import stopwords


In [28]:
# Importing cell

import string
import re
from collections import Counter
from nltk import ngrams

In [7]:
# Gutenborg corpora accessing

infile1 = open('/content/drive/My Drive/LGCS124/Copy of austen-emma.txt', 'r')
austen_emma = infile1.read()
infile2 = open('/content/drive/My Drive/LGCS124/Copy of austen-sense.txt', 'r')
austen_sense = infile2.read()
infile3 = open('/content/drive/My Drive/LGCS124/Copy of austen-persuasion.txt', 'r')
austen_persuasion = infile3.read()
infile4 = open('/content/drive/My Drive/LGCS124/Copy of bible-kjv.txt', 'r')
bible_kjv = infile4.read()
infile5 = open('/content/drive/My Drive/LGCS124/Copy of shakespeare.txt', 'r')
shakespeare = infile5.read()
infile6 = open('/content/drive/My Drive/LGCS124/Copy of whitman-leaves.txt', 'r')
whitman_leaves = infile6.read()
infile7 = open('/content/drive/My Drive/LGCS124/Copy of independence.txt', 'r')
independence = infile7.read()
infile8 = open('/content/drive/My Drive/LGCS124/Copy of milton-paradise.txt', 'r')
milton_paradise = infile8.read()
infile9 = open('/content/drive/My Drive/LGCS124/Copy of hawthorne-gables.txt', 'r')
hawthorne_gables = infile9.read()
infile10 = open('/content/drive/My Drive/LGCS124/Copy of constitution.txt', 'r')
constitution = infile10.read()
infile11 = open('/content/drive/My Drive/LGCS124/Copy of blake-songs.txt', 'r')
blake_songs = infile11.read()
infile12 = open('/content/drive/My Drive/LGCS124/Copy of blake-poems.txt', 'r')
blake_poems = infile12.read()
infile13 = open('/content/drive/My Drive/LGCS124/Copy of dickens-oliver.txt', 'r')
dickens_oliver = infile13.read()
infile14 = open('/content/drive/My Drive/LGCS124/Copy of dickens-expectations.txt', 'r')
dickens_expectations = infile14.read()
infile15 = open('/content/drive/My Drive/LGCS124/Copy of dickens-drood.txt', 'r')
dickens_drood = infile15.read()

### 1. Only one author in /data/corpora/Gutenberg uses the word 'and' more often than 'the'. Who is it? (Hint: It's ok to simplify this question by looking only at lower-case uses of these two words)

In [ ]:
# Lowercasing all the texts

list_corpus = [austen_emma, austen_sense, austen_persuasion, bible_kjv, 
               shakespeare, whitman_leaves, independence, milton_paradise, 
               hawthorne_gables, constitution, blake_songs, blake_poems, 
               dickens_oliver, dickens_expectations, dickens_drood]

def lower_all(my_list):  # Convert all texts to lowercase and store them as vars lower+number
  for i in range(0,15):
    globals()['lower' + str(i)] = my_list[i].lower()  
  return globals()['lower' + str(i)]

lower_all(list_corpus)
#print(lower1[:1000])

In [9]:
# Function to find all 'and' and return the number of occurances

def count_and(text):
  and_tokens = re.findall(r'\s+and\s+', text)
  return len(and_tokens)

# Func to find all 'the'
def count_the(text):
  the_tokens = re.findall(r'\s+the\s+', text)
  return len(the_tokens)


In [14]:
# Find them in all texts

and_list = [] # initialize
the_list = []

# Generate two lists containing the number of and and the in all texts
for i in range(0,15):
  and_list.append(count_and(globals()['lower' + str(i)]))
  the_list.append(count_the(globals()['lower' + str(i)]))

# Tell me which author in the list!
def tell_and_the():
  for i in range(0,15):
    if and_list[i] > the_list[i]:
      return i
  return None

tell_and_the()

7

Therefore, the text that is the 7th (counting from 0), which is `milton_paradise`, contains more 'and' than 'the'!

### 2. How many word types appear in the Bible exactly twice?


In [24]:
# Continue using lowercase

lower_bib = lower3 # Just so that I don't have to remember its sequence number

tokens_bib = re.findall(r'[a-z]+', lower_bib)  # Tokenizes the lowercased bible

count_bib = list(Counter(tokens_bib).items())  # Make the counter into a list of tuples

# A function that returns the list of words that appear two times
def two_times():
  two_times_list = []
  for word, times in count_bib:  # Since I have already made the word counts into a list of 2-tuples
    if times == 2:
      two_times_list.append(word)
  return two_times_list

print(len(two_times())) # Print out the length of the list above

1743


Therefore, 1743 words appear exactly twice in the Bible!

### 3. How many of the word types (vocabulary items) in the Declaration of Independence never occur in Shakespeare??

In [25]:
# Lowercase, strip punctuations, count tokens


independence = independence.lower()
independence = independence.translate(str.maketrans('', '', string.punctuation))
ind_tokens = nltk.tokenize.word_tokenize(independence)
ind_c = Counter(ind_tokens)


shakespeare = shakespeare.lower()
shakespeare = shakespeare.translate(str.maketrans('', '', string.punctuation))
sha_tokens = nltk.tokenize.word_tokenize(shakespeare)
sha_c = Counter(sha_tokens)


# Print the length of the set that contains elements included in independence but not shakespeare

print(len(set(ind_c.keys()) - set(sha_c.keys())))

103


Therefore, 103 word types only appear in `independence` but not `shakespeare`!


### 4. What are the three most frequent TRIgrams in Blake’s poems??

In [30]:
blakep_lower = lower11  # Just so that I don't have to remember its sequence number

tokens_blakep = re.findall(r'[a-z]+', blakep_lower)  # Tokenizes the lowercased bible

bp3grams = list(ngrams(tokens_blakep, 3))  # Creating a list of trigrams

trigram_counter = Counter(bp3grams)  # Counts trigrams

print(trigram_counter.most_common(3))  # Most common 3

[(('the', 'voice', 'of'), 5), (('the', 'human', 'form'), 5), (('can', 'it', 'be'), 5)]


Thus the most frequent trigrams are as shown above!

### 5. Take a look at “About Ngram Viewer” (https://books.google.com/ngrams/info), especially advanced options for searches.

Use Google Ngram Viewer (books.google.com/ngrams) to look at a diachronic
(across time) effect related in some way to language of prejudice, discrimination, or otherwise related to a historically oppressed population. You might look at something to do with race, ethnicity, religion, class, gender, gender identification, sexual orientation, age, ableism... Produce a VERY short write-up, just a few sentences is ok: (1) what phenomenon did you look at; (2) your search parameters (terms, range
of years, other options); (3) a capture of your resulting graph; (4) very briefly, what you think the graph reveals (Did it match or not match your expectations?).

1) I am looking at the relative representation of gay-identifying men versus women.

2) Search parameters: 
  - Keyword: `gay women + lesbian + gay woman,gay men + gay + gay man`. This might be slightly flawed as sometimes the term 'gay' is slightly broader than 'lesbian' as women sometimes so identify without specifing 'gay women', so it might lead to an unproportionately high occurance of the latter Ngram.
  - Range of years: 1960-2019. The reason is that the term *gay* has multiple meanings and is especially common in old literature to express happiness instead of the meaning I am looking for, so there was a significant peak around 1850-1920 that I would not like to look at.

3) 

![image](https://user-images.githubusercontent.com/77285010/153723772-31f0a7b7-1b57-451f-8ba7-568c1412c1ab.png)

4) It did match my expectation that within the underrepresented group of gay people, men are still significantly more represented than women, perhaps due to the traditional taboo attached to the discussion of the desire, preferences and intimate identities of women. What is surprising is that the occurance of the term is even declining in recent years -- I wonder what has happened.